### DISCUSSION TITLE
How do I re-use assigned ports?

### DISCUSSION THREAD

**Discussion by NikiforovAll** [microsoft/aspire#4539] [Q&A] [answered]
Context:

As a developer, I want to be able to configure my tooling and use it between application restarts.

E.g: I configured tooling to interact with a database and want to use it between application restarts. How would I achieve that? 

Currently, I constantly check the content of connection strings and logs to find corresponding services. The workflow is somewhat tedious and interrupts me from the "flow".

**Comment by davidfowl** [comment]
You can set ports and passwords manually. That's the best option when you are using external tools. The Aspire.Hosting.* packages should have methods that let you specify the port and password (if they generate one on the fly).

**Comment by NikiforovAll** [comment]
`WithPgAdmin` doesn't provide this option. Should it be fixed? 

```csharp
var usersDb = builder
    .AddPostgres("dbserver", pAdmin, password)
    .WithDataVolume()
    .WithPgAdmin()
    .AddDatabase("users-db");
```

**Reply by davidfowl** [reply, marked as answer]
```C#
var usersDb = builder
    .AddPostgres("dbserver", pAdmin, password)
    .WithDataVolume()
    .WithPgAdmin(c => c.WithHostPort(8001))
    .AddDatabase("users-db");
```

**Comment by NikiforovAll** [comment]
From my point of view, it would be great to persist randomly assigned ports between sessions without assigning values explicitly in AppHost. 

```csharp
var builder = DistributedApplication.CreateBuilder(args).WithPersistentExternalPorts();
```

**Reply by davidfowl** [reply]
Not sure, that requires state. Launch profiles are visible state that can be tweaked, understood, and changed by users. I think we'd need something similar here. Not a bad idea overall but I'm not a fan of this method.

**Reply by johan-krs** [reply]
Reviving this - now the proxy exists, it might make the ask cleaner. 

The proxy knows the random host port (it routes to it), but `WithReference` injects the proxy's own random listener port into consumers, so connection strings churn on every aspire start even though nothing about the container has moved.

Setting `IsProxied = false` currently ends up using the well-known TargetPort for resources like Postgres/SQL Server/Redis, which collides with locally-installed services on those ports. Could this be changed?

Could `WithReference` inject the container's actual host port instead of the proxy's? 
That would give: 
- random port on first create
- stable across restarts
- reachable externally while the AppHost is stopped
- and no clashes with local installs 
 
Without removing the proxy for cases that genuinely need it (replicas, project resources).

**Reply by davidfowl** [reply]
Open an issue. 

cc @danegsta

**Reply by danegsta** [reply]
For containers in 13.4 we've made a few changes to how proxyless endpoints work and we've made proxyless the default for persistent containers. Proxyless container endpoints get either the hardcoded fixed public port or a random port assigned by the runtime. That's already in the daily builds if you want to try them out, otherwise it won't be long before 13.4 ships.
